### DISCUSSION TITLE
Email queue spontaneously switched off

### DISCUSSION THREAD

**Discussion by StickyPixels** [liedekef/events-made-easy#449] [Q&A]
Hi,
I've had this issue recently on two different sites now, so I'm guessing there is a bug somewhere: Email queueing gets randomly switched off for no obvious reason. Emails suddenly stop being sent, and the schedule is switched off alltogether in Email Settings.
I checked and in both cases there has been no interaction that may have caused the issue (like plugin updates or maintenance or anything like that).
It appears to be a recent thing, because one of these sites has been active for years and I never had this issue before. I am using the default Wordpress Email method on both sites.

I found this post in this forum: [https://github.com/liedekef/events-made-easy/discussions/331](url), may this be CRON related? 

Hope there is a solution or workaround, but anything to clarify would help.

**Comment by liedekef** [comment]
The EME option to set queueing to active is calleeme_queue_mailsd "eme_queue_mails". The only place where EME code touches the setting is in the EME options page, tab "Email". And again only if you're in that tab, EME sets/unsets the option based on standard wordpress functions (in fact the wordpress options API takes care of it). *And* the default is even "1" (meaning "on"). So if something is setting this to 0, it is not an EME bug but something else doing/causing this. If it happens, check your WP database, table wp-options, and search for that option name/value. Maybe it got deleted (and then it is most probably a WP bug).
Edit: this reply is written based on the info in the ticket (no plugin upgrades, no maintenance). Also, no info is given on the used EME version, so this is about all I can say.

**Comment by StickyPixels** [comment]
Thank you for clarifying, I checked my database but not sure what to look for. Anyway I reset the option earlier so probably the option has been re-added.
Do you think the problem won't happen again if I set another email method, like php email or smtp?

**Reply by liedekef** [reply]
it's totally unrelated to the email method setting so ...

**Comment by roysinn** [comment]
I've had this same problem a few times. Thought it might be connected with updating EME but just happened today or yesterday and I have not updated EME (still running 2.5.20). Like StickyPixels, just can't connect it to anything I'm doing or not doing with EME. I just reschedule in Settings and wait for somebody to scream whenever they haven't received an email. Frequency - not often, maybe every 4-6 months. Sorry - not very helpful but if I find something more definitive, I'll post it.

**Reply by liedekef** [reply]
And which option is turned off in the Settings then? Is it "Use mail queuing?" or "Send out queued mails in batches of" ?

**Reply by roysinn** [reply]
Previous and desired setting was "Email queue settings | Send out queued mails in batches of 10 | Once every minute (EME schedule)"
Magically changed to "Email queue settings | Send out queued mails in batches of 10 | Not scheduled"

So, maintained the batch size but schedule was changed. 

By the bye, I have another site using EME and it was updated to 2.5.22 and then again today to 2.5.24. No problem. Unchanged.

**Reply by roysinn** [reply]
Thanks Franky. In all cases, "Use email queuing" is set to "Yes" and is not reset. Nor the batch size. Just seems the schedule setting is spontaneously changing to "Not scheduled". I'm sure I don't reset it but I recall doing something in the EME Settings recently on the site where it got reset.

**Reply by liedekef** [reply]
Well, there is a long outstanding issue in wordpress where planned actions get removed. It is very logical too in fact: they store all planned actions in 1 option, meaning actions can overwrite others because it gets updated after each run ... 
Currently this is the ticket: https://core.trac.wordpress.org/ticket/57271 where disappearing schedules are discussed.
Anyway, if it is set to 1 minute, then also check your wordpress php logs for "could not reschedule" error. Maybe use the 5-minute schedule, and also don't use WP cron but system cron.

**Comment by roysinn** [comment]
Thanks for the suggestions, Franky. You've added to the voices that suggest using the system cron instead of the WP cron. 
If I do find anything more definitive, I'll share. In the meantime, I can manage the things by checking the settings more often.

**Reply by liedekef** [reply]
Well, system cron isn't really the problem (but setting things at 1 minute is kind of cutting things close, it is also the shortest interval for system cron). The real issue is WP cron implementation. I've added my suggested solution to the mentioned thread, but I'm not ready to implement this for WP that soon :-)

**Comment by f14m07** [comment]
Hi,
Is it possible to set the EME send queue function in system cron different from the other functions programmed in wp-cron? ?

**Reply by liedekef** [reply]
I'm not sure I understand your question ... do you mean another method of calling EME send queue without relying on WP cron functionality?

**Comment by f14m07** [comment]
On my Cron System I have planned 
```
wget -q -O - https://vogue-et-reve.fr/wp-cron.php >/dev/null 2>&1
```
as you suggest in the doc. 
I was wondering if it would be possible to add something like 
```
wget -q -O - https://vogue-et-reve.fr/EME-function.php >/dev/null 2>&1
```

**Comment by f14m07** [comment]
Today I am using this hook:
```php
/* Contrôle du CRON */
add_action( 'eme_insert_rsvp_action', 'mon_test_cron' );
add_action('eme_insert_member_action','mon_test_cron' );
add_action('eme_insert_event_action','mon_test_cron' );
add_action('emefs_submit_event_action','mon_test_cron' );

function mon_test_cron() {
   if (!wp_next_scheduled('eme_cron_send_queued')){
    wp_schedule_event(time(), 'eme_1min', 'eme_cron_send_queued');
	$current_schedule = wp_get_schedule( 'eme_cron_send_queued' );
    wp_mail('frederic.moriniere@gmail.com','CRON reprogrammé','eme_cron_send_queued reprogrammé, le temps est : '.$current_schedule);
}
}
/*end Contrôle du CRON */
```
I can add other add_action or maybe add this in my system CRON?

**Reply by liedekef** [reply]
Well, that's an ugly hack :-)
Anyway, you forced me to step back and take a look at things, sooooo .... I added a REST API call so you can call that from cron to process the queue. 
From the changelog:
```
* Added a REST API call, in case you don't trust WP cron to process the queue. As an example, you can call:
  curl --insecure --user "username:password" https://localhost/wordpress/wp-json/events-made-easy/v1/processqueue/60
  ==> don't user --insecure for public sites, and change "https://localhost/wordpress" by your wordpress url
  ==> change the "username" by your user and the "password" by an application password generated in your WP user settings
  ==> "60" means the script can run at most for 55 seconds (=60-5, 5 being a safety measure). Never set this higher than your cron recurrence of course
  ==> set the timing option for queue processing to "not scheduled" in EME, so the two don't interfere :-) But in fact: it doesn't real
ly matter: EME is resilient enough to cope with both at the same time (but better be safe than sorry).
```
The change is in the current dev version.

**Reply by f14m07** [reply]
To be sure I have well understood. I must put this line "curl ..." as I did for "wget ..." in the system CRON of my hosting?

**Reply by liedekef** [reply]
yes, the wget line stays in cron (as it was before: it triggers wordpress cron via system cron)
The curl line can be wget too (as long as you use the correct options to pass username/password), but if you aren't familiar just do as indicated in the changelog and use curl. And feel free to add ">/dev/null 2>&1" too if you don't care about possible command output.

**Reply by f14m07** [reply]
Ok. Thanks for the work. 
I'd rather wait for the next version to test this and then I'll give you some feedback.

**Reply by liedekef** [reply]
I've already tested this locally and it seems to work as expected :-) But of course it's up to you to decide this. I don't expect the version to be released this year though, I just released one yesterday. And I'm pretty proud on my jtable-release too :-D

**Reply by f14m07** [reply]
If I upload the three files cli_mail, eme-cron and eme-actions, is it suffisant ?

**Comment by liedekef** [comment]
The relevant files are eme-actions, eme-cron, eme-mailer and eme-options.

**Reply by f14m07** [reply]
What I did:
- I upload the files on my Local web site. 
- I create the password/user app. 
- I disabled 'Use mail queuing'.
- I am not sure how make system cron on a Local setting. So I went on the shell site and I entered the curl command.
What I observed:
- The use mail queuing setting is still No
- On the schedule actions tab, it's written "Queued mails will be send out in batches of 30 Once every minute (EME schedule)" but alsao "Email queueing is not activated.
Because mail queueing is not activated, the newsletter functionality is not available."
- The mails are sent immediately.

What I did wrong?

**Reply by liedekef** [reply]
You switched off mail queueing, then of course mails are sent immediately. You need to leave mail queueing active and disable the schedule in EME (so  instead of "once every xxx minutes/hours" you choose "not scheduled"):
![image](https://github.com/user-attachments/assets/b06c8da2-f887-48f3-96c3-0d5c1f36b1cf)

**Reply by f14m07** [reply]
Ok. I did it.
I have tested mails.
I have know this text in the schedule actions tab : "There are 2 messages in the mail queue.
WARNING: messages found in the queue but the mail queue is not configured correctly, so they will not be sent out via WP. Make sure to use the registered REST API call with appropriate options from system cron to process the queue". The news letter seems Ok.
As I don't know haw to set the curl command in the system CRON, do i need to put the curl command in the shell site at any time?

**Reply by liedekef** [reply]
Put it in system cron with the same frequency as you did with the wp-cron.php call ...
The message in the "scheduled actions" tab still need some refining to accomodate this new possibility :-)

**Comment by f14m07** [comment]
Hi,
This evening I've set up the new task in the CRON system (on the production site), set up the dev version of the 4 files, and set up the queue.
The test emails worked well. But I'm still surprised because you never see the mails in the queue. It would be nice if the "schedule actions/Email queue info" was also update in real time. 
Similarly, a scheduled mailing is not saved in the mailings tab.
Is it possible?

Thanks a lot for the work

**Reply by liedekef** [reply]
that's because you still have the dev-code :-)
This is now my printout in schedule actions/Email queue info":

```
Email queue info
There are 3 messages in the mail queue.
WARNING: messages found in the queue but the mail queue is not scheduled to be processed by WP, so they will not be sent out via WP. Make sure to use the registered REST API call with appropriate options from system cron to process the queue.

```
I think I'll change the text later on a bit ...

**Reply by liedekef** [reply]
btw: the new release is out.
